### PR TITLE
update/0.3.0

### PR DESCRIPTION
**Pull request by decanus**

### CODE REVIEW

File: Sources/BeaconChain/DataStructures/State/Validator.swift [hunk 1]
```diff
@@ -5,11 +5,11 @@ struct Validator {
...
-    let withdrawalEpoch: UInt64
-    var penalizedEpoch: UInt64
+    var withdrawableEpoch: UInt64
+    var slashedEpoch: UInt64
     var statusFlags: UInt64
     
-    func isActive(epoch: EpochNumber) -> Bool {
-        return self.activationEpoch <= epoch && epoch < self.exitEpoch
+    func isActive(epoch: Epoch) -> Bool {
+        return activationEpoch <= epoch && epoch < exitEpoch
```
Comment (djrtwo) [inline Sources/BeaconChain/DataStructures/State/Validator.swift]: Fancy. Don't need `self.`?

File: Sources/BeaconChain/StateTransition.swift [hunk 2]
```diff
@@ -254,17 +254,17 @@ extension StateTransition {
     }
 
     static func exits(state: inout BeaconState, block: BeaconBlock) {
```
Comment (djrtwo) [inline Sources/BeaconChain/StateTransition.swift]: Might consider renaming `voluntary_exits`

File: Sources/BeaconChain/StateTransition.swift [hunk 3]
```diff
@@ -280,6 +280,49 @@ extension StateTransition {
...
+            assert(state.validatorBalances[Int(transfer.from)] >= transfer.amount)
+            assert(state.validatorBalances[Int(transfer.from)] >= transfer.fee)
+            assert(
+                state.validatorBalances[Int(transfer.from)] == transfer.amount + transfer.fee
+                || state.validatorBalances[Int(transfer.from)] >= transfer.amount + transfer.fee + MIN_DEPOSIT_AMOUNT
+            )
+
+            assert(state.slot == transfer.slot)
+            assert(BeaconChain.getCurrentEpoch(state: state) >= state.validatorRegistry[Int(transfer.from)].withdrawableEpoch)
+            // @todo does this work (-1)?
```
Comment (djrtwo) [inline Sources/BeaconChain/StateTransition.swift]: Does it?

Comment (djrtwo) [reply Sources/BeaconChain/StateTransition.swift]: If it does work, wouldn't it be `1 ..<= -1` so you include the last element?

File: Sources/BeaconChain/StateTransition.swift [hunk 4]
```diff
@@ -297,7 +340,7 @@ extension StateTransition {
 extension StateTransition {
 
     static func processEpoch(state: inout BeaconState) {
-        assert(state.slot + 1 % EPOCH_LENGTH == 0) // @todo not sure if this should be here
+        assert(state.slot + 1 % SLOTS_PER_EPOCH == 0) // @todo not sure if this should be here
```
Comment (djrtwo) [inline Sources/BeaconChain/StateTransition.swift]: You should only call this function if that condition holds. Seems a little redundant to assert it here but not necessarily incorrect

File: Sources/BeaconChain/DataStructures/State/Validator.swift (hunk 1 above)
Comment (decanus) [reply Sources/BeaconChain/DataStructures/State/Validator.swift]: Nope, art.

File: Sources/BeaconChain/StateTransition.swift (hunk 3 above)
Comment (decanus) [reply Sources/BeaconChain/StateTransition.swift]: @djrtwo found something even nicer: https://developer.apple.com/documentation/foundation/data/1779685-suffix

**Pull request merged**